### DISCUSSION TITLE
StreamProvider releasing all of streams over the whole app

### DISCUSSION THREAD

**Discussion by KristianBalaj** [rrousselGit/riverpod#1134] [Q&A]
I encountered this challenge at multiple projects overall.

Imagine using snapshots with Firebase (streaming changes).

In case there are security rules that a user can listen to this entity only when logged in, then in case of logout all these streams finish with error. And in case I have error logging in the app, I don't want to log errors over the whole app when a user logs out 😄 

I wonder if there is any available mechanism with riverpod to solve this artifact? 
I'm thinking about any possibility to close all streams that `StreamProvider`s hold. 
I think there is some container with all the available providers (`ProviderScope`), maybe there is a possibility to iterate over all the providers check if it is a `StreamProvider` and close it somehow?

What is a nice approach to solve this problem? Does riverpod provide any clean solution?

**Comment by rrousselGit** [comment]
You could make a separate provider that watches whether the user is logged in or not.

Then within the providers listening to collection snapshots, watch this new provider. This way when the user logs out, the collection snapshot providers will refresh.

**Comment by KristianBalaj** [comment]
Aggreed, that looks like a solution.

But there are following problems to that:

1. When the user logs out, the streams already throw errors, because then the user cannot even close the stream (it ends with error because of missing auth - user is logged out)
2. I could have an another provider e.g. `final loggingOutProvider = StateProvider<bool>((_) => false)`, I set this provider to true right before I log out the user and to true when the user logs in. This provider will make all the `StreamProvider`s return `Stream.empty` in all the places where the Firebase stream should be closed. However, I think this throws, too [NOT SURE]. Because it maybe does not closes the streams immediatelly, or does it? The problem could be that it closes the stream after some delay when the user is already signed out, then the close is unauthenticated and also throws.

**I'm not sure if the 2. point could happen. What do you think?**

Also a problem with this approach is that I could forget to add that logic into every relevant provider 🤔 
Could this be achieved somewhat cleaner with riverpod? 🤔

**Reply by rrousselGit** [reply]
The fact that the previous streams will throw doesn't matter, because the error will be caught by the provider and your UI should see the loading state instead
